R/readWrite.R

```diff
@@ -0,0 +1,96 @@
+    write.table(transcript_counts, file= transcript_countsfn, sep="\t",quote=FALSE)
+    gene_se <- transcriptToGeneExpression(se)
+    gene_counts <- as.data.frame(assays(gene_se)$counts)
+    gene_countsfn <- paste(path,"counts_gene.txt",sep="/")
+    write.table(gene_counts, file= gene_countsfn, sep="\t", quote=FALSE)
+    }
+  return (gtf)
+}
+#' Outputs a GTF file for the nanorna-bam nextflow pipeline 
+#' @title transcript to gene expression
+#' @param grList a GRangesList object
+#' @param file the output gtf file name
+#' @param geneIDs an optional dataframe of geneIDs (column 2) with the corresponding transcriptIDs (column 1)
+#' @return gtf a GTF dataframe
+#' @export
+write.gtf <- function (grList,file,geneIDs) {
+  if (missing(grList) | missing(file)){
+    stop('Both GRangesList and the name of file are required.')
+  }else if (class(grList) != "CompressedGRangesList"){
+    stop('The inputted GRangesList is of the wrong class.')
+  }else {
+    df <- as.data.frame(grList)
+    df$exon_rank <- paste('exon_number "',df$exon_rank,'";',sep= '')
+    if (missing(geneIDs)){
+      df$group_name <- paste('transcript_id "',df$group_name,'";',sep= '')
+      gtf_exon <- data.frame(seqname=df$seqnames, source= "Bambu",feature= "exon",
+                             start=df$start,end=df$end,score=".",strand=df$strand,frame=".",
+                             attributes= paste(df$group_name,df$exon_rank))
+    }else{
+      df <- merge(df,geneIDs,by.x="group_name",by.y="TXNAME",all=TRUE)
+      df$group_name <- paste('transcript_id "',df$group_name,'";',sep= '')
+      df$GENEID <- paste('gene_id "',df$GENEID,'";',sep= '')
+      gtf_exon <- data.frame(seqname=df$seqnames, source= "Bambu",feature= "exon",
+                             start=df$start,end=df$end,score=".",strand=df$strand,frame=".",
+                             attributes= paste(df$GENEID,df$group_name,df$exon_rank))
+    }
+    df_end <- df[order(df$end, decreasing = TRUE),]
+    df_end <- data.frame(group_name=df_end$group_name,uend=df_end$end)
+    df_end <- df_end[!duplicated(df_end$group_name),]
+    df <- df[!duplicated(df$group_name),]
+    df <- merge(df,df_end,by="group_name",all=TRUE)
+    if (missing(geneIDs)){
+      gtf_trns <- data.frame(seqname=df$seqnames, source= "Bambu",feature= "transcript",
+                             start=df$start,end=df$uend,score=".",strand=df$strand,frame=".",
+                             attributes= paste(df$group_name))
+    }else{
+      gtf_trns <- data.frame(seqname=df$seqnames, source= "Bambu",feature= "transcript",
+                             start=df$start,end=df$uend,score=".",strand=df$strand,frame=".",
+                             attributes= paste(df$GENEID,df$group_name))
+    }
+    gtf <- rbind(gtf_trns,gtf_exon)
+    gtf <- gtf[order(gtf$attributes),]
+  } 
+  write.table(gtf, file= file, quote=FALSE, row.names=FALSE, col.names=FALSE, sep = "\t")
+  return (gtf)
```

Comment:
the function does not need to return the gtf object

Reply:
The function does not return a GTF dataframe now. I have also removed the test_that for the GTF file format.